neqo-transport/src/pmtud.rs

```diff
@@ -0,0 +1,644 @@
+    /// times, the PMTUD process is stopped.
+    pub fn on_packets_lost(
+        &mut self,
+        lost_packets: &[SentPacket],
+        stats: &mut Stats,
+        now: Instant,
+    ) {
+        if lost_packets.is_empty() {
+            return;
+        }
+
+        let mut increase = vec![0; self.search_table.len()];
+        let mut loss_counts_updated = false;
+        for p in lost_packets {
+            let idx = self
+                .search_table
+                .iter()
+                .take_while(|&&sz| p.len() > sz - self.header_size)
+                .count();
+            // Count each lost packet size <= the current MTU only once. Otherwise a burst loss of
+            // >= MAX_PROBES MTU-sized packets triggers a PMTUD restart. Counting only one of them
+            // here requires three consecutive loss instances of such sizes to trigger a PMTUD
+            // restart.
+            //
+            // Also, ignore losses of packets <= the minimum QUIC packet size, (`searchtable[0]`),
+            // since they just increase loss counts across the board, adding to spurious
+            // PMTUD restarts.
+            if idx > 0 && (increase[idx] == 0 || p.len() > self.plpmtu()) {
+                loss_counts_updated = true;
+                increase[idx] += 1;
+            }
+        }
+
+        if !loss_counts_updated {
+            return;
+        }
+
+        let mut accum = 0;
+        for (c, incr) in zip(&mut self.loss_counts, increase) {
+            accum += incr;
+            *c += accum;
+        }
+
+        // Track lost probes
+        let lost = self.count_pmtud_probes(lost_packets);
+        stats.pmtud_lost += lost;
+
+        // Check if any packet sizes have been lost MAX_PROBES times or more.
+        //
+        // TODO: It's not clear that MAX_PROBES is the right number for losses of packets that
+        // aren't PMTU probes. We might want to be more conservative, to avoid spurious PMTUD
+        // restarts.
+        let Some(first_failed) = self.loss_counts.iter().position(|&c| c >= MAX_PROBES) else {
+            // If not, keep going.
+            if lost > 0 {
+                // Don't stop the PMTUD process.
+                self.probe_state = Probe::Needed;
+            }
+            return;
+        };
+
+        if first_failed > 0 {
+            let last_ok = first_failed - 1;
+            qdebug!(
+                "Packet of size > {} lost >= {} times",
+                self.search_table[last_ok],
+                MAX_PROBES
+            );
+            if self.probe_state == Probe::NotNeeded {
+                // We saw multiple losses of packets <= the current MTU outside of PMTU discovery,
+                // so we need to probe again. To limit connectivity disruptions, we start the PMTU
+                // discovery from the smallest packet up, rather than the failed packet size down.
+                self.restart_pmtud(stats);
```

Comment:
Want to add a TODO here:

If we are declaring losses, that means that we'll be getting packets through.  The size of those will put a floor on the MTU.  We're currently conservative and starting from scratch, but we don't strictly need to do that.

(If our loss recovery processing deals with lost packets before acked packets, maybe this ends up being a no-op, I haven't checked that.  That said, if it IS a no-op, we should explain that here and maybe write some tests in support of that.  Hence the suggestion for a TODO.)